jb/itch5/compute_base.cpp

```
@@ -0,0 +1,193 @@
+}
+
+void jb::itch5::compute_base::handle_reduce(
+    time_point recv_ts, long msgcnt, std::size_t msgoffset,
+    message_header const& header, std::uint64_t order_reference_number,
+    std::uint32_t shares, bool all_shares) {
+  auto result = handle_reduce_no_update(
+      recv_ts, msgcnt, msgoffset, header, order_reference_number,
+      shares, all_shares);
+  check_callback_condition(recv_ts, header, result);
+}
+
+jb::itch5::compute_base::update_result
+jb::itch5::compute_base::handle_reduce_no_update(
+    time_point recv_ts, long msgcnt, std::size_t msgoffset,
+    message_header const& header, std::uint64_t order_reference_number,
+    std::uint32_t shares, bool all_shares) {
+  // First we need to find the order ...
+  auto position = orders_.find(order_reference_number);
+  if (position == orders_.end()) {
+    // ... ooops, this should not happen, there is a problem with the
+    // feed, log the problem and skip the message ...
+    JB_LOG(warning) << "missing order id=" << order_reference_number
+                    << ", location=" << msgcnt << ":" << msgoffset
+                    << ", header=" << header;
+    return std::make_tuple(false, order_data(), books_.end());
+  }
+  // ... okay, now that the order is located, find the book for that
+  // symbol ...
+  auto& data = position->second;
+  auto stock_book_it = books_.find(data.stock);
+  if (stock_book_it == books_.end()) {
+    // ... ooops, this should not happen, there is a problem with the
+    // book... an order existed (position) but there is no book for that symbol
+    // ...log the problem and skip the message ...
+    JB_LOG(warning) << "missing book for symbol id: " << data.stock 
+                    << ", order id=" << order_reference_number
+                    << ", location=" << msgcnt << ":" << msgoffset
+                    << ", header=" << header;
+    return std::make_tuple(false, order_data(), books_.end());    // error= books_.end()
```

should we throw instead of returning an error in this case?  Just a question, please consider it, make a decision and move on.

---

throw it will be! I will fix it for each error (warning) condition